src/misc/unicode.cpp

```diff
@@ -0,0 +1,1129 @@
+        mark_2nd) { // can't add more than 2 combining marks, not supported
+
+        Invalidate();
+        return;
+    }
+
+    const uint8_t mark = static_cast<uint8_t>(code_point - combining_first + 1);
+
+    if (mark_1st)
+        mark_2nd = mark;
+    else
+        mark_1st = mark;
+}
+
+void Grapheme::StripMarks()
+{
+    mark_1st = 0;
+    mark_2nd = 0;
+}
+
+bool Grapheme::operator<(const Grapheme &other) const
+{
+    // TODO: after migrating to C++20 use a templated lambda here
+
+    if (code_point < other.code_point)
+        return true;
+    else if (code_point > other.code_point)
+        return false;
+
+    if (mark_1st < other.mark_1st)
+        return true;
+    else if (mark_1st > other.mark_1st)
+        return false;
+
+    if (mark_2nd < other.mark_2nd)
+        return true;
+    else if (mark_2nd > other.mark_2nd)
+        return false;
+
+    if (empty < other.empty)
+        return true;
+    else if (empty > other.empty)
+        return false;
+
+    if (valid < other.valid)
+        return true;
+    else if (valid > other.valid)
+        return false;
+
+    return false;
+}
+
+// ***************************************************************************
+// Conversion routines
+// ***************************************************************************
+
+static uint8_t char_to_uint8(const char in_char)
+{
+    if (in_char >= 0)
+        return static_cast<uint8_t>(in_char);
+    else
+        return static_cast<uint8_t>(UINT8_MAX + in_char + 1);
+}
+
+static char uint8_to_char(const uint8_t in_uint8)
+{
+    if (in_uint8 <= INT8_MAX)
+        return static_cast<char>(in_uint8);
+    else
+        return static_cast<char>(in_uint8 - INT8_MIN + INT8_MIN);
+}
+
+static void utf8_to_wide(const std::string &str_in, std::vector<uint16_t> &str_out)
+{
+    // Convert UTF-8 string to a sequence of decoded integers
+
+    // For UTF-8 encoding explanation see here:
+    // - https://www.codeproject.com/Articles/38242/Reading-UTF-8-with-C-streams
+    // - https://en.wikipedia.org/wiki/UTF-8#Encoding
+
+    str_out.clear();
+    for (size_t i = 0; i < str_in.size(); i++) {
+        const size_t  remaining = str_in.size() - i - 1;
+        const uint8_t byte_1 = char_to_uint8(str_in[i]);
+        const uint8_t byte_2 = (remaining >= 1) ? char_to_uint8(str_in[i + 1]) : 0;
+        const uint8_t byte_3 = (remaining >= 2) ? char_to_uint8(str_in[i + 2]) : 0;
+
+        // Retrieve code point
+        uint32_t code_point = UINT32_MAX; // dummy, to trigger unsupported
+        if (byte_1 < 0x80) {
+
+            code_point = byte_1; // 1-byte code point, ASCII compatible
+
+        } else if (GCC_UNLIKELY(byte_1 < 0xc0)) {
+
+            code_point = unknown_character; // not an UTF-8 code point
+
+        } else if (GCC_UNLIKELY(byte_1 >= 0xfc)) {
+
+            i += 5; // 6-byte code point (>= 31 bits), no support
+
+        } else if (GCC_UNLIKELY(byte_1 >= 0xf8)) {
```

Comment:
How about extracting all these checks into well-named lambdas, then you could get rid of most these explanatory comments?

Reply:
Lambdas are probably an overkill here - after turning the magic values into constexpr's it should now be clear enough.